### PR TITLE
Libraries: Blobstore

### PR DESCRIPTION
**Pull request by bori00**
The user can attach a photo to the comment that they want to send

The recent changes rely on the Blobstore API, which enables uploading and storing blobs. I added a feature to the comments functionality, which allows users to attach a photo to the comment that they want to send. The action triggered by submitting the form is set to be the url generated by blobstore. When blobstore finishes processing the photo, it redirects the request to the comments-data servlet, as before. This servlet, additional to what has been implemented until now, gets the url generated by blobstore for the image, and stores in in the database. Thus, when the user wants to see the comments on the comments.html page, the messages can be displayed together with the photos, using the urls as image sources.

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/comment/Comment.java [hunk 1]
```diff
@@ -9,16 +9,20 @@
 import com.google.appengine.api.datastore.PreparedQuery;
 import com.google.appengine.api.datastore.Query;
 import com.google.appengine.api.datastore.FetchOptions;
+import org.jetbrains.annotations.Nullable;
 
 
 /** Stores the data related to one comment */
```
Comment (tlheng) [inline portfolio/src/main/java/com/google/sps/comment/Comment.java]: nits: ends with period `.`

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java [hunk 2]
```diff
@@ -89,7 +104,7 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
         ArrayList<CommentData> result = new ArrayList<CommentData>();
         for(Comment comment : comments) {
```
Comment (tlheng) [inline portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java]: fix spaces

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java [hunk 3]
```diff
@@ -89,7 +104,7 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
         ArrayList<CommentData> result = new ArrayList<CommentData>();
```
Comment (tlheng) [inline portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java]: I think Java8 allows omit type: ArrayList<CommentData> result = new ArrayList<>();

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java [hunk 4]
```diff
@@ -115,11 +130,43 @@ private User getUserFromRequest(HttpServletRequest request){
     }
 
     private Comment getCommentFromRequest(String senderId, HttpServletRequest request) {
-        String message = request.getParameter("comment");
-        Comment newComment = new Comment(senderId, message);
+        String message = request.getParameter(COMMENT_INPUT_NAME);
+        @Nullable String imgUrl = getUploadedFileUrl(request, IMG_INPUT_NAME);
```
Comment (tlheng) [inline portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java]: I don't think it is necessary to annotate local variable.

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java [hunk 5]
```diff
@@ -115,11 +130,43 @@ private User getUserFromRequest(HttpServletRequest request){
...
+        System.out.println("!!!The image's url is: " + imgUrl);
```
Comment (tlheng) [inline portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java]: Is this needed?

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java [hunk 6]
```diff
@@ -115,11 +130,43 @@ private User getUserFromRequest(HttpServletRequest request){
...
+            return null;
+        }
+
+        ImagesService imagesService = ImagesServiceFactory.getImagesService();
+        ServingUrlOptions options = ServingUrlOptions.Builder.withBlobKey(blobKey);
+        try {
+            URL url = new URL(imagesService.getServingUrl(options));
+            return url.getPath();
+        } catch (MalformedURLException e) {
+            return imagesService.getServingUrl(options);
```
Comment (tlheng) [inline portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java]: what will this return?

File: portfolio/src/main/webapp/comments.js [hunk 7]
```diff
@@ -17,23 +17,54 @@ function displayComments(){
     .catch(error => console.log('failed to print comments to DOM: ' + error));
 }
 
-function deleteAllComments(){
+function addAllCommentsToDOM(commentDatas) {
+    console.log("add all comments to DOM: "+  commentDatas.length)
+    for(var i=0; i<commentDatas.length; i++) {
```
Comment (tlheng) [inline portfolio/src/main/webapp/comments.js]: fix spaces

File: portfolio/src/main/webapp/comments.js [hunk 8]
```diff
@@ -17,23 +17,54 @@ function displayComments(){
...
     var cardHolder = document.createElement("div");
     cardHolder.setAttribute('class', 'card-holder-small');
     return cardHolder;
 }
 
-function createCardElement(highlight){
+function createCardElement(highlight) {
     var card = document.createElement("div");
     card.setAttribute('class', 'card-small');
     if(highlight === true){
```
Comment (tlheng) [inline portfolio/src/main/webapp/comments.js]: fix spaces

File: portfolio/src/main/java/com/google/sps/comment/repository/impl/DatastoreCommentRepository.java [hunk 9]
```diff
@@ -30,6 +31,7 @@ private Entity getCommentEntity(Comment comment) {
         Entity commentEntity = new Entity(ENTITY_NAME);
         commentEntity.setProperty(MESSAGE_PROPERTY, comment.getMessage());
         commentEntity.setProperty(SENDER_ID_PROPERTY, comment.getSenderId());
+        commentEntity.setProperty(IMG_URL_PROPERTY, comment.getImgUrl());
```
Comment (tlheng) [inline portfolio/src/main/java/com/google/sps/comment/repository/impl/DatastoreCommentRepository.java:34]: 4 spaces indentation? I think we need to pick if we want 2 or 4 spaces.

**Review by tlheng — APPROVED** [review]
Nice cl

File: portfolio/src/main/java/com/google/sps/comment/Comment.java (hunk 1 above)
Comment (bori00) [reply portfolio/src/main/java/com/google/sps/comment/Comment.java]: done

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java (hunk 2 above)
Comment (bori00) [reply portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java]: done

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java (hunk 3 above)
Comment (bori00) [reply portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java]: changed

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java (hunk 4 above)
Comment (bori00) [reply portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java]: removed annotation

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java (hunk 5 above)
Comment (bori00) [reply portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java]: removed print statement

File: portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java (hunk 6 above)
Comment (bori00) [reply portfolio/src/main/java/com/google/sps/servlets/CommentsServlet.java]: changed to string

File: portfolio/src/main/webapp/comments.js (hunk 7 above)
Comment (bori00) [reply portfolio/src/main/webapp/comments.js]: done

File: portfolio/src/main/webapp/comments.js (hunk 8 above)
Comment (bori00) [reply portfolio/src/main/webapp/comments.js]: done

File: portfolio/src/main/java/com/google/sps/comment/repository/impl/DatastoreCommentRepository.java (hunk 9 above)
Comment (bori00) [reply portfolio/src/main/java/com/google/sps/comment/repository/impl/DatastoreCommentRepository.java:34]: From what I can see, it's constant 4. I'm not sure about what I need to change here.

**Pull request merged**